src/librustc_mir/build/matches/mod.rs

```diff
@@ -398,17 +407,41 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
             candidates.iter().take_while(|c| c.match_pairs.is_empty()).count();
         debug!("match_candidates: {:?} candidates fully matched", fully_matched);
         let mut unmatched_candidates = candidates.split_off(fully_matched);
-        for (index, candidate) in candidates.into_iter().enumerate() {
+
+        let fully_matched_with_guard =
+            candidates.iter().take_while(|c| c.guard.is_some()).count();
+
+        let unreachable_candidates = if fully_matched_with_guard + 1 < candidates.len() {
+            candidates.split_off(fully_matched_with_guard + 1)
+        } else {
+            vec![]
+        };
+
+        for candidate in candidates {
             // If so, apply any bindings, test the guard (if any), and
             // branch to the arm.
-            let is_last = index == fully_matched - 1;
-            if let Some(b) = self.bind_and_guard_matched_candidate(block, arm_blocks,
-                                                                   candidate, is_last) {
+            if let Some(b) = self.bind_and_guard_matched_candidate(block, arm_blocks, candidate,
+                                                                   opt_imaginary_target_block) {
                 block = b;
             } else {
                 // if None is returned, then any remaining candidates
                 // are unreachable (at least not through this path).
-                return vec![];
+                // Link them with false edges.
+                debug!("match_candidates: add false edges for unreachable {:?} and unmatched {:?}",
+                       unreachable_candidates, unmatched_candidates);
+                let dummy_source_info = self.source_info(DUMMY_SP);
+                for candidate in unreachable_candidates {
```

Comment:
can't you iterate over `unreachable_candidates.zip(unmatched_candidates)` in the for-loop here? These are both unreachable candidates, and should be treated similarly. I think the current way will cause an infinite loop if you have a partial candidate after a full candidate, e.g.:
```Rust
match Some(42) {
    x => {},
    Some(y) if maybe => {},
    z => {}
}
```

In any case, you should add a mir opt test for that,